feat: update of the Fast Data tutorial (#380)
mia-platform/documentation: docs/tutorial/fast_data/fast_data_tutorial.mdx

| **registry_id** | String | The registry ID. Check this field as "Required" and "Primary Key".|
| **name**        | String | The name of the customer to whom the registry refers.             |
| **email**       | String | The email of the customer to whom the registry refers.            |
| **phone**       | String | The phone of the customer to whom the registry refers.            |
| **address**     | String | The address of the customer to whom the registry refers.          |
| **city**        | String | The city of the customer to whom the registry refers.             |
| **state**       | String | The State of the customer to whom the registry refers.            |

#### Ingestion Topics

At the bottom of the same page, we can also see the Kafka topics for each environment. There are two types of topics (_Ingestion topics_ and _PR Update topics_) 
but within the scope of this tutorial, only the ingestion topics are considered. The Kafka Ingestion topics will be used by the Real Time Updater 
to receive the data from the Source Systems so that the Projections will be updated, but we will discuss this later.

> ![Create System of Records](img/05b-page-overview-projection-kafka.png) 

Author: This might be an hard take, but I prefer to not change the topic names (for now) because I'm assuming that this tutorial might be in a dedicated demo project where the topics are already configured.

Anyway, rn the topics are automatically generated as: company, environment, system, projection name, suffix.
I'd prefer to keep it that way.

| **order_id**    | String | The order ID. Check this field as "Required" and "Primary Key".|
| **name**        | String | The name of the product.                                       |
| **price**       | Number | The price of the product.                                      |
| **date**        | Date   | The date when the order was placed.                            |
| **registry_id** | String | The registry ID of the customer who placed the order.          |

#### Ingestion Topics

These are the topics defined in Kafka where we expect to receive messages related to the update of projections. The following are the suggested names for the Ingestion topics:

- **demo.development.delivery.order.ingestion**: the topic that will be used to send the data to the order projection in the _development_ environment
- **demo.integration.delivery.order.ingestion**: the topic that will be used to send the data to the order projection in the _integration_ environment
- **demo.production.delivery.order.ingestion**: the topic that will be used to send the data to the order projection in the _production_ environment

### Microservice Realtime Updater Configuration

> Now that we have one System of Records with two Projections, we [save the configuration](/tutorial/console/console_handbook.md#save-changes) until now. 

Reviewer: ```suggestion
Now that we have one System of Records with two Projections, we [save the configuration](/tutorial/console/console_handbook.md#save-changes).
```

This service will automatically be named _fast-data-delivery-realtime-updater_ and it can be found in the "Microservices" section.

![Real Time Updater Microservice](img/09-real-time-updater-microservice.png)

Now we need to configure all the parameters of the microservice. Here is an example of the fields that we need to configure:

- **PROJECTIONS_DATABASE_NAME**: the name of the database that will be used to store the data of the Projections
- **PROJECTIONS_CHANGES_COLLECTION_NAME**: the name of the collection that will be used to store the changes of the Projections (we suggest _fd-pc-delivery_)
- **KAFKA_BROKERS**: the list of the kafka brokers that will be used to send the data to the Projections
- **KAFKA_GROUP_ID**: the group id that will be used to consume the data from the source systems
- **KAFKA_SASL_USERNAME**: the username that will be used to login to the kafka brokers
- **KAFKA_SASL_PASSWORD**: the password that will be used to login to the kafka brokers 
- **KAFKA_SASL_MECHANISM**: the mechanism that will be used to login to the kafka brokers

:::info

> The Real-Time Updater includes more optional parameters that migth be included via Environment Variables. If you need to know more about them, you can [read the documentation of the Real-Time Updater](/fast_data/configuration/realtime_updater.md).

Reviewer 2: ```suggestion
The Real-Time Updater includes more optional parameters that might be included via Environment Variables. If you need to know more about them, you can [read the documentation of the Real-Time Updater](/fast_data/configuration/realtime_updater.md).
```

But in the example above, we decided to configure the relationship both ways to be able to use the same JSON also in the service that aggregates the Single View (the _Single View Creator_).
</TabItem>

<TabItem value="no-code">
:::info
The following explains how to configure an ER Schema using the "No Code" feature. To do so, please make sure that this feature is enabled in your project. You might need to ask your Project Administrator for it.
:::

To configure the ER Schema with the No Code, you need to return to the "Projection" section, and then select the "ER Schemas" tab. From there, you can create a new ER Schema, by clicking the "Create new ER Schema" button:
a modal will prompt you to choose a name for your ER Schema (we suggest: _delivery-er-schema_). After that, you will be redirected to the ER Schema canvas.

![Create new ER Schema](img/11-create-er-schema-no-code.png)

The configuration of the ER Schema with the "No Code" is explained in detail in the [dedicated section](/fast_data/configuration/config_maps/erSchema.md#use-the-no-code), 
but we include here a quick guide on how to create the ER Schema with the needed relationships:

> - In the right side panel you can see the list of Projections inside the System of Record _delivery_: darg the collections _registry_ and _order_ to the canvas in the middle of the screen.

Reviewer: ```suggestion
- In the right side panel you can see the list of Projections inside the System of Record _delivery_: drag the collections _registry_ and _order_ to the canvas in the middle of the screen.
```

which means that the relationship has been created but the condition inside is empty.
- the right side panel should show the details of the relationship. From here you'll need to:
  - change the "Condition Type" value of "A -> B" to _One to many_.
  - click on the "Edit Rules" button to open the Rules Modal: click on "+ Rule" and then fill the input fields with "registry_id" both for the _registry_ and the _order_ Projection. Finally, click the "Save" button.

The final result should resemble the following image:

![The ER Schema configuration is completed](img/12-completed-er-schema-no-code.png)

After that, the ER Schema is already completely configured. To be used in the Real-Time Updater microservice, however, you need to copy the JSON code (visible by clicking on the "\<\/\>" button at the top-left side of the canvas)
and paste it in the ER Schema ConfigMap of the microservice (from the "Microservices" section, select the Real-Time Updater named _fast-data-delivery-realtime-updater_, 
go to the "ConfigMaps & Secret" and select the "er-schema" tab).

</TabItem>

</Tabs>

#### Projection Changes Schema

> The _Projection Changes Schema_ is a ConfigMap that represents the path that goes from one entity to another entity, and it is used to find the Single View key. 

Reviewer: ```suggestion
The _Projection Changes Schema_ is a ConfigMap that represents the path that goes from one entity to another, and it is used to find the Single View key. 
```

In this tutorial, our Single View is called _order-history_ and we will create an object schema with the registry information and the orders information linked to the registry.

### Microservice Single View Creator Configuration

First of all, we need to create a microservice that we will attach to the Single View that we're going to create. This microservice is called "Single View Creator", and it's available in the microservice Marketplace.
![Marketplace and Single View Creator](img/23-marketplace-single-view-low-code.png)

Here we have two different plugins: the first one is for the Single View Creator, and the second one is for the Single View Creator Low Code. 
We will go for the latter since we want to leverage the usage of ConfigMaps for a quicker configuration of the microservice. We suggest naming it _single-view-creator-delivery_.

Now we need to configure all the parameters of the microservice, here is an example of the fields that we need to configure:

> - **PROJECTIONS_CHANGES_COLLECTION**: it is the name of the collection that includes the Projection Changes generated by the Real-Time Updater, the two must share the same name (we previously suggested _fd-pc-delivery_).

Reviewer: `the two must share the same name` what does `the two` refer to? The two projections?

Author: let me know if you like it more this way

- **SINGLE_VIEWS_COLLECTION**: should be the name of the single view which your single view creator is responsible for (we suggest _order-history_, the same name as the Single View we are going to create).
- **SINGLE_VIEWS_ERRORS_COLLECTION**: it is the name of a MongoDB CRUD you want to use as a collection for the single view errors (we suggest: _order-history-errors_).
- **SINGLE_VIEWS_PORTFOLIO_ORIGIN**: should be equal to the SYSTEM_ID you have set in PROJECTIONS_CHANGES_COLLECTION (if you follow our name conventions, you will set _delivery_)
- **TYPE**: identifies the type of projection changes that need to be read. It should be the same as the Single View name you want to update (also here we are going to call it _order-history_).

:::info

> Also the Single View Creator includes more optional parameters that migth be included via Environment Variables. If you need to know more about them, you can [refer to the related documentation section](/fast_data/configuration/single_view_creator.md#environment-variables).

Reviewer: ```suggestion
The Single View Creator includes more optional parameters that might be included via Environment Variables. If you need to know more about them, you can [refer to the related documentation section](/fast_data/configuration/single_view_creator.md#environment-variables).
```

:::

The microservice also needs the following ConfigMaps to work properly:
  * **Single View Key**: this is the key that will be used to store the Single View
  * **ER-Schema**: the ER Schema (should be the same as the one used in the Projection Microservice)
  * **Aggregation**: the aggregation that will be used to calculate the Single View

But we are going to update these later, inside the Single View page.

### Single View "order-history"

> Inside our Single View, each document is an object that contains the registry information about the user, and the information about the orders associated with the user.

Reviewer: ```suggestion
Inside our Single View, each document is an object that contains the registry information about the user and the information about the orders associated with the user.
```
